// @wordpress/components/build-module/menu/item.js
/**
 * WordPress dependencies
 */
import { forwardRef, useContext } from '@wordpress/element';

/**
 * Internal dependencies
 */

import * as Styled from './styles';
import { MenuContext } from './context';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
export const MenuItem = forwardRef(function MenuItem({
  prefix,
  suffix,
  children,
  hideOnClick = true,
  ...props
}, ref) {
  const menuContext = useContext(MenuContext);
  if (!menuContext?.store) {
    throw new Error('Menu.Item can only be rendered inside a Menu component');
  }
  return /*#__PURE__*/_jsxs(Styled.MenuItem, {
    ref: ref,
    ...props,
    accessibleWhenDisabled: true,
    hideOnClick: hideOnClick,
    store: menuContext.store,
    children: [/*#__PURE__*/_jsx(Styled.ItemPrefixWrapper, {
      children: prefix
    }), /*#__PURE__*/_jsxs(Styled.MenuItemContentWrapper, {
      children: [/*#__PURE__*/_jsx(Styled.MenuItemChildrenWrapper, {
        children: children
      }), suffix && /*#__PURE__*/_jsx(Styled.ItemSuffixWrapper, {
        children: suffix
      })]
    })]
  });
});
